Source/DataSources/RelativeToTerrainHeightProperty.js

```diff
@@ -0,0 +1,201 @@
+            get : function() {
+                return this._positions;
+            },
+            set : function(value) {
+                var oldValue = this._positions;
+                if (oldValue !== value) {
+                    if (defined(oldValue)) {
+                        this._subscription();
+                    }
+
+                    this._positions = value;
+
+                    if (defined(value)) {
+                        if (!value.isConstant) {
+                            throw new RuntimeError('positions must be a constant property');
+                        }
+                        this._subscription = value._definitionChanged.addEventListener(function() {
+                            this._fetchTerrainHeight();
+                            this._definitionChanged.raiseEvent(this);
+                        }, this);
+                    }
+
+                    this._fetchTerrainHeight();
+                    this._definitionChanged.raiseEvent(this);
+                }
+            }
+        },
+        heightRelativeToTerrain : createPropertyDescriptor('heightRelativeToTerrain')
+    });
+
+    var centroidScratch = new Cartesian3();
+    function computeCentroid(positions) {
+        var centroid = Cartesian3.clone(Cartesian3.ZERO, centroidScratch);
+        var length = positions.length;
+        for (var i = 0; i < length; i++) {
+            centroid = Cartesian3.add(positions[i], centroid, centroid);
+        }
+        return Cartesian3.multiplyByScalar(centroid, 1/length, centroid);
+    }
+
+    /**
+     * @private
+     */
+    RelativeToTerrainHeightProperty.prototype._fetchTerrainHeight = function() {
+        this._terrainHeight = 0;
+        var property = this._positions;
+        if (!defined(property)) {
+            return;
+        }
+        var positions = property.getValue(Iso8601.MINIMUM_VALUE);
+        if (!defined(positions)) {
+            return;
+        }
+        if (!isArray(positions)) {
+            positions = positions.positions; //positions is a PolygonHierarchy, just use the outer ring
+        }
+        var centroid = computeCentroid(positions);
+        var carto = Cartographic.fromCartesian(centroid, undefined, this._cartographicPosition);
+        carto.height = 0.0;
+        var that = this;
+        sampleTerrainMostDetailed(this._terrainProvider, [carto])
+            .then(function(results) {
+                if (that._positions !== property) {
+                    return;
+                }
+                that._terrainHeight = defaultValue(results[0].height, 0);
+                that._definitionChanged.raiseEvent(that);
+            });
+    };
+
+    /**
+     * Gets the height relative to the terrain based on the positions.
+     *
+     * @returns {Number} The height relative to terrain
+     */
+    RelativeToTerrainHeightProperty.prototype.getValue = function(time) {
+        return this._terrainHeight + this.heightRelativeToTerrain.getValue(time);
```

Comment:
Always use `Property.getValue` when accessing values in our code.  Probably `Property.getValueOrDefault` in this case to avoid NaN